common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java

```diff
@@ -194,53 +344,98 @@ private void handleSmartcardCertAuth(@NonNull final ClientCertRequest request) {
                     //Create List that contains cert details only pertinent to the cert picker.
                     final List<YubiKitCertDetails> certList = getCertDetailsFromKey(piv);
                     //If no certs were found, cancel flow.
                     if (certList.isEmpty()) {
-                        request.cancel();
                         Logger.info(methodTag,  "No PIV certificates found on YubiKey device.");
-                        //TODO: show error dialog stating that no PIV certificates were found on the YubiKey.
+                        //Show error dialog stating that no PIV certificates were found on the YubiKey.
+                        dialogHolder.showErrorDialog(R.string.smartcard_no_cert_dialog_title, R.string.smartcard_no_cert_dialog_message);
+                        request.cancel();
                         return;
                     }
                     //Build and show Smartcard cert picker, which also handles the rest of the smartcard CBA flow.
-                    showSmartcardCertPickerDialog(certList, request);
-
-                } catch(final IOException e) {
-                    Logger.error(methodTag, e.getMessage(), e);
-                    request.cancel();
-                    //TODO: create and show error dialog here.
-                } catch (final ApduException e) {
-                    Logger.error(methodTag, e.getMessage(), e);
-                    request.cancel();
-                    //TODO: create and show error dialog here.
-                } catch (final ApplicationNotAvailableException e) {
+                    dialogHolder.showCertPickerDialog(
+                            certList,
+                            getSmartcardCertPickerDialogPositiveButtonListener(request),
+                            new SmartcardCertPickerDialog.NegativeButtonListener() {
+                                @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
+                                @Override
+                                public void onClick() {
+                                    request.cancel();
+                                    dialogHolder.dismissDialog();
+                                }
+                            },
+                            new SmartcardCertPickerDialog.CancelCbaCallback() {
+                                @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
+                                @Override
+                                public void onCancel() {
+                                    request.cancel();
+                                }
+                            });
+                } catch(final IOException | ApduException | BadResponseException e) {
                     Logger.error(methodTag, e.getMessage(), e);
+                    //Show general error dialog.
+                    dialogHolder.showErrorDialog(R.string.smartcard_general_error_dialog_title, R.string.smartcard_general_error_dialog_message);
                     request.cancel();
```

Comment:
Does the order of showing dialog and canceling the request matter?

Reply:
I don't think so, but I tried to keep consistent (show dialog, then cancel). I'll check again and make sure I'm consistent here.